### DISCUSSION TITLE
`Tensor` to `ndarray`: surrogates.utils.sequential_posterior - `candidates` shape and reshaping

### DISCUSSION THREAD

**Discussion by rjavadi** [emdgroup/baybe#154] [Q&A] [answered]
I've converted the type of `candidates` to a NumPy ndarray. However, this change introduced an issue with using `flatten()`. NumPy's `flatten()` method doesn't have `end_dim` argument. Therefore, I need to find an alternative approach for flattening/reshaping the array while preserving the desired structure.

Any clues or suggestions regarding this refactoring is appreciated.

**Comment by AVHopp** [comment]
Can you point us to the point where exactly this changed should happen?

**Reply by AVHopp** [reply]
Ahhh I should maybe also read the title of a conversation :D

**Comment by Scienfitz** [comment]
@rjavadi what are you aiming to achieve with this? the surrogate modeul seems close to the bayesian optimziation and hence it seems should use tensors

**Comment by rjavadi** [comment]
@Scienfitz In our meetings we discussed that the only surrogate model that requires using tensors is gaussian process, so I changed the base class to use ndarray instead of torch.

I may be wrong :) it's also less work for me if we only need to load torch lazily here 😅

**Reply by Scienfitz** [reply, marked as answer]
hmm I'm not sure that was correct, I think its more like: anything that is used inside `acquisition` can keep using torch. While the surrogates are strictly speaking not derived of acquisition, they are only ever needed if acquisition is also used so I guess it would be fine to keep them having torch (lazy)

Re your original question:
Did you try with reshape?
ie a ndarray of shape `[2,4,3,4]` would normally be flattened to `[96]` but you could just flatten everything except the last dimensions (according to what `end_dim`  was chosen), for example here `end_dim=-2` ignoring the last dimension resulting in shape `[24, 4]`. I would hope this creates the right order

We need to keep in mind the performance of the code in the gaussian process code, ie it might be that numpy performs copies where torch doesn't or similar so that it would be slower

**Reply by rjavadi** [reply]
Thanks for reply! I'm going to revert `surrogates` changes and use torch.

**Comment by Scienfitz** [comment]
closing as no longer relevant
